### ISSUE TITLE
[KModal]: Investigate KModal backdrop entrance transition not being smooth anymore

### ISSUE THREAD

**Issue by AlexVelezLl** [learningequality/kolibri-design-system#848]
## Product
I have found this issue while testing https://github.com/learningequality/studio/pull/4842 that updates Studio to the latest KDS release. And I have barely noticed it in Kolibri Too

## Expected behavior
KModal backdrop entrance transition should be smooth as it was in 4.x.x. Example video of using KDS 4.5.1:

https://github.com/user-attachments/assets/191c550b-c780-44d2-b4db-e3fbfdd27789

## Actual behavior

KModal backdrop now in KDS 5.0.0-rc10 shows a choppy animation.

https://github.com/user-attachments/assets/127573f1-a6e5-45fb-8eb6-7c40f6410ce2

Where even during some frames, only some html elements are affected by the backdrop and others are not.

![Image](https://github.com/user-attachments/assets/e905bef8-e7b4-43a9-8974-55c844c4da89)

## Steps to reproduce the issue
1. Check out this Studio PR: https://github.com/learningequality/studio/pull/4842
2. Open a modal
3. Check the KModal entrance animation.

**Comment by MisRob** [comment]
- [Guidance on previewing KDS in Studio](https://github.com/learningequality/kolibri-design-system/blob/develop/dev_docs/03_how_to_update_library.md#2-preview-updates-in-a-product)
- First step would be to figure out the cause. Then we can discuss possible solutions.

**Comment by Sahil-Sinha-11** [comment]
Hey @MisRob, I would like to work on this issue, please assign it to me.

**Comment by MisRob** [comment]
Hi @Sahil-Sinha-11, sure thanks! I am not sure if this is better task then https://github.com/learningequality/kolibri-design-system/issues/690, usually 'investigate' tasks can be tricky. However I am happy if we can find out more about what's the cause or at least some more information, it's still very helpful :)

Since this bug is specific to Studio, a possible culprit may be some conflicts with Vuetify styles (one example https://github.com/learningequality/kolibri-design-system/pull/851) so perhaps that'd be first thing to check on.

**Comment by MisRob** [comment]
Hi @Sahil-Sinha-11, sure thanks! I am not sure if this is better task than https://github.com/learningequality/kolibri-design-system/issues/690, usually 'investigate' tasks can be tricky. However I am happy if we can find out more about what's the cause or at least some more information, it's still very helpful :)

**Comment by Sahil-Sinha-11** [comment]
Hey @MisRob @AlexVelezLl, I linked the KDS with Kolibri, but I’m not seeing the changes reflected yet. Would you mind helping me figure out what might be going wrong? I’d really appreciate your assistance! Thanks.

https://github.com/user-attachments/assets/4f296bcd-fc6e-464d-a1b8-22ad1b22eca3

**Comment by AlexVelezLl** [comment]
Hi @Sahil-Sinha-11! After updating the KDS codebase you need to wait a little bit for the js bundler to build your changes again, can you please make sure that webpack is rebuilding the plugins after you update kds?

**Comment by Sahil-Sinha-11** [comment]
Hey @AlexVelezLl, yes, I did wait for a few seconds and this is what it shows... It is detecting changes but not reflecting on the local host , what should I do?
![Image](https://github.com/user-attachments/assets/067bd1ca-100c-4c24-82c8-575dea0b4b30)

**Comment by AlexVelezLl** [comment]
Could you check your Kolibri node_modules folder and see if the `kolibri-design-system` package is soft linked?

**Comment by Sahil-Sinha-11** [comment]
> Could you check your Kolibri node_modules folder and see if the `kolibri-design-system` package is soft linked?

![Image](https://github.com/user-attachments/assets/efa697b5-dc55-4ac1-ae3d-cfbebd454f5c)
 this one?

**Comment by AlexVelezLl** [comment]
Yes, I see it is indeed soft linked. Hmmm maybe is your browser caching the js files? It only happens with KDS changes? Can you try opening the page in incognito mode?

**Comment by Sahil-Sinha-11** [comment]
> Yes, I see it is indeed soft linked. Hmmm maybe is your browser caching the js files? It only happens with KDS changes? Can you try opening the page in incognito mode?

I tried on incognito but it is still the same . what should I do?

**Comment by AlexVelezLl** [comment]
Can you please try stopping the devserver, clearnig the webpack cache (removing the `node_modules/.cache/webpack` folder), then restart the devserver, and do a hard refresh on the page?

Can you also check if this only happens with KDS changes?

**Comment by Sahil-Sinha-11** [comment]
Thanks a lot, @AlexVelezLl After I cleared the node_modules/.cache/webpack it works now.

**Comment by Sahil-Sinha-11** [comment]
Hey @AlexVelezLl, while investigating I changed the transition time to be 1 second and I found the transition happens only on the page not on the appbar, once the transition is complete on page then it goes to appbar, I am not sure but maybe it has something to do with koverlay, am I in the right direction?
https://github.com/user-attachments/assets/e8b8836d-587f-44d9-9e55-d8698dca7822

**Comment by AlexVelezLl** [comment]
Yes, its a good direction. You should probably also see what are the differences between KModals in v4.0.0 (release-v0.17.x branch in Kolibri, and release-v4 branch in KDS) and v5.0.0 (develop branches on Kolibri and KDS), and play with both of them seeing the differences.

**Comment by Sahil-Sinha-11** [comment]
Hey @AlexVelezLl, after my research I think the problem might be in
1. The function getOverlayEl relies on window.overlayEl in the _useoverlay, which is efficient after initialization. However, if initialization (mountOverlay) fails or is delayed so when the transition begins it might not have been mounted.
2.  The frequent movement of elements between the DOM and the overlay container using KOverlay and Teleport might be causing problem.

Although I reverted changes made in kmodal in the previous version v4.0.0 in kds I didn't find any difference on screen. How can I proceed further from here?

**Comment by MisRob** [comment]
If it's useful for confirming if the issue originated with the overlay logic, this is my pull request that introduced it https://github.com/learningequality/kolibri-design-system/pull/722

**Comment by Sahil-Sinha-11** [comment]
Hey @MisRob , i was trying to mount the toolbar in the way below but I did not observed any changes am I doing it in the right way ? 
<img width="1350" alt="Image" src="https://github.com/user-attachments/assets/b53a27e4-cbe7-44d3-9e76-2ece42deced6" />

**Comment by MisRob** [comment]
@Sahil-Sinha-11 what are you trying to achieve or experiment with here? This script is meant to attach the overlay element to DOM in products such as Kolibri or Studio when [KDS is installed](https://design-system.learningequality.org/installation/#install-plugin), so I don't exactly understand.

**Comment by MisRob** [comment]
@Sahil-Sinha-11 Also Sahil, I wanted to let you know that will be offline starting later today and note that [we will be closed from December 23 to January 5](https://github.com/learningequality/kolibri/discussions/12956).

**Comment by Sahil-Sinha-11** [comment]
> [@Sahil-Sinha-11](https://github.com/Sahil-Sinha-11) what are you trying to achieve or experiment with here? This script is meant to attach the overlay element to DOM in products such as Kolibri or Studio when [KDS is installed](https://design-system.learningequality.org/installation/#install-plugin), so I don't exactly understand.

I am trying to mount the appbar inside the overlay when created here
<img width="1350" alt="Image" src="https://github.com/user-attachments/assets/e7bc992e-49e9-4f63-9855-c6590fb4054d" />

**Comment by MisRob** [comment]
@Sahil-Sinha-11 I don't think KDS should be mounting `KToolbar` automatically upon initialization. `KToolbar` is only a component about which consumers decide whether it should be used. That's different from the overlay element. It would help me to understand why you are trying to do this - would you describe the general purpose of the experimentation?

**Comment by MisRob** [comment]
@Sahil-Sinha-11 Also Sahil, I wanted to let you know that will be offline starting later today and note that [we will be closed from December 23 to January 5](https://github.com/learningequality/kolibri/discussions/12956).

**Comment by Sahil-Sinha-11** [comment]
> [@Sahil-Sinha-11](https://github.com/Sahil-Sinha-11) Also Sahil, I wanted to let you know that will be offline starting later today and note that [we will be closed from December 23 to January 5](https://github.com/learningequality/kolibri/discussions/12956).

Thank you for letting me know, @MisRob  I appreciate the heads-up. Looking forward to working on more interesting issues together when you're back. Have a wonderful holiday break!

**Comment by Sahil-Sinha-11** [comment]
> [@Sahil-Sinha-11](https://github.com/Sahil-Sinha-11) I don't think KDS should be mounting `KToolbar` automatically upon initialization. `KToolbar` is only a component about which consumers decide whether it should be used. That's different from the overlay element. It would help me to understand why you are trying to do this - would you describe the general purpose of the experimentation?

while Investigating I found the toolbar wasn't reflecting the transition as expected so I thought it was not mounted inside the overlay so the transition delay was happening, so I tried to mount it,
<img width="1710" alt="Image" src="https://github.com/user-attachments/assets/57ece338-5c7c-4716-89cd-b067e9b3a60d" />
when I removed it the transition worked fine on page.

**Comment by MisRob** [comment]
@Sahil-Sinha-11 Ah alright, no the toolbar is not mounted inside the overlay, that's intended. Basically no elements are mounted there when Kolibri is initialized, we teleport modals to the overlay element later. We couldn't have the toolbar in the overlay in any case, because modals need to be on top of everything, toolbars included - so I don't think we could possible resolve this issue by doing something like this.

**Comment by MisRob** [comment]
@Sahil-Sinha-11 Looking at all this and particularly [this comment](https://github.com/learningequality/kolibri-design-system/issues/848#issuecomment-2542320636), it seems that the timing is the main culprit, right? Here are some thought about what I would try to look at if I were investigating:
- (1) What would happen if you delayed `KModal` transition slightly, would it help?
- (2) You could also try to replace [this KModal Vue `transition`](https://github.com/learningequality/kolibri-design-system/blob/develop/lib/KModal.vue#L5-L8) with native [CSS transition](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_transitions)
- (3) Combination of both

These are really just experimental ideas that could give us further insight. If that doesn't bear any fruit, I think we could leave it for now unless someone else has another ideas - I don't know what would I do myself at this point. We may need to re-consider how to improve our approach to teleporting overall. In any case, all the things you've found out so far are already very valuable :) Thanks!

**Comment by MisRob** [comment]
@Sahil-Sinha-11 Oh and one more thing to try out may be to see what would happen if you removed the Vue `transition` from inside the modal to outside of it, just temporarily to see if it has any effect

```
<transition...
  <KModal...>
```

These ideas may not necessarily be solutions but if you investigated in detail if/how/why it helped with the timing issues, we could understand even a bit better what would be possible direction to deal with this.

**Comment by Sahil-Sinha-11** [comment]
> [@Sahil-Sinha-11](https://github.com/Sahil-Sinha-11) Looking at all this and particularly [this comment](https://github.com/learningequality/kolibri-design-system/issues/848#issuecomment-2542320636), it seems that the timing is the main culprit, right? Here are some thought about what I would try to look at if I were investigating:
> 
> * (1) What would happen if you delayed `KModal` transition slightly, would it help?
> * (2) You could also try to replace [this KModal Vue `transition`](https://github.com/learningequality/kolibri-design-system/blob/develop/lib/KModal.vue#L5-L8) with native [CSS transition](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_transitions)
> * (3) Combination of both
> 
> These are really just experimental ideas that could give us further insight. If that doesn't bear any fruit, I think we could leave it for now unless someone else has another ideas - I don't know what would I do myself at this point. We may need to re-consider how to improve our approach to teleporting overall. In any case, all the things you've found out so far are already very valuable :) Thanks!

I tried the 1st point by changing the time of transition to 1 sec, I observed that first the transition happens on the page except ktoolbar, when it finishes on the page the transition happens on the toolbar immediately.

**Comment by Sahil-Sinha-11** [comment]
I’ll give the ideas a try you shared and see how they work. Thanks for the suggestions!

**Comment by MisRob** [comment]
@Sahil-Sinha-11 Ah okay, so it seems you think it could be something specific to `KToolbar`. Then you might try to gradually remove some markup or styles from it, perhaps, to get closer to the cause - it's one of the common debugging techniques. Alright, thanks and see you later in January!

**Comment by Sahil-Sinha-11** [comment]
Hey @MisRob @AlexVelezLl, I found the bug If I remove the KFocusTrap the transition is smooth.
Before removing

https://github.com/user-attachments/assets/356f2c8c-f6e8-4dc0-814a-951680868ea6

After removing

https://github.com/user-attachments/assets/b5e634fe-5295-4712-91ea-77d20602493b

**Comment by Sahil-Sinha-11** [comment]
@MisRob I saw that KFocudTrap was added to prevent accidental interaction  with components outside modal, Is there any specific area where this issue arises? I didn't see this problem in some places where I opened a modal.

**Comment by marcellamaki** [comment]
Hi @Sahil-Sinha-11 - thanks for your work on this. The `KFocusTrap` is our implementation of the general principle of ["focus trapping" which is essential for proper accessibility](https://okenlabs.com/blog/accessibility-implementing-focus-traps/). So, we need to find a solution that resolves the transition smoothness without removing `KFocusTrap`.

**Comment by AlexVelezLl** [comment]
Thats a very good insight @Sahil-Sinha-11!! If removing KFocusTrap solved it, then perhaps the problem could be the `<transition>` tag that wraps the KFocusTrap not being applied correctly. Could you investigate further about that? Perhaps switching the order of these two elements (the `KFocusTrap`, and the `transition`) could be the solution. But could you explore what are the implications of this? and determine if its safe to do it?

**Comment by Sahil-Sinha-11** [comment]
> Thats a very good insight [@Sahil-Sinha-11](https://github.com/Sahil-Sinha-11)!! If removing KFocusTrap solved it, then perhaps the problem could be the `<transition>` tag that wraps the KFocusTrap not being applied correctly. Could you investigate further about that? Perhaps switching the order of these two elements (the `KFocusTrap`, and the `transition`) could be the solution. But could you explore what are the implications of this? and determine if its safe to do it?

After switching it worked but how do I check its safety?

**Comment by AlexVelezLl** [comment]
You can look at why the switching works. What do the transition tag do that doesnt work well with the KFocustrap, but it does with the child div, for example.

**Comment by MisRob** [comment]
@AlexVelezLl @Sahil-Sinha-11 If it seems that the Vue `transition` is the main culprit, we may also consider using native CSS transition instead. In some nuanced cases this approach can work better. Feel free to experiment @Sahil-Sinha-11. Good progress!

**Comment by Sahil-Sinha-11** [comment]
@MisRob @AlexVelezLl  I tried to use normal css like this:-

<img width="425" alt="Image" src="https://github.com/user-attachments/assets/3ead31cb-2071-48d2-9416-90f144f1abcb" />

<img width="425" alt="Image" src="https://github.com/user-attachments/assets/deef7315-cf74-4831-ba4f-dc7a899ca85a" />

but it didn't work.Also I tried to wrap div inside kfocustrap but didn't get expected result.  
 Would there be any problem If we just switch the KfocusTrap and Transition ?

**Comment by MisRob** [comment]
@Sahil-Sinha-11 

> Would there be any problem If we just switch the KfocusTrap and Transition ?

I can't think of anything now. If it helps, let's give it a try. Then you can see how it looks like.

Side note, I don't know if the code snippets are full but if yes, than the implementation is not complete since you're not saying what style is transitioning. More [here](https://developer.mozilla.org/en-US/docs/Web/CSS/transition). But no need to do this if the other approach works.

**Comment by Sahil-Sinha-11** [comment]
> Hey [@MisRob](https://github.com/MisRob) [@AlexVelezLl](https://github.com/AlexVelezLl), I found the bug If I remove the KFocusTrap the transition is smooth. Before removing
> 
>  Screen.Recording.2025-01-06.at.5.35.48.PM.mov 
> After removing
> 
>  Screen.Recording.2025-01-06.at.5.34.59.PM.mov

Hey @MisRob @AlexVelezLl,it works like this after switching should I raise a PR?

**Comment by AlexVelezLl** [comment]
Thanks @Sahil-Sinha-11. Yes, please! :)

**Comment by Sahil-Sinha-11** [comment]
@AlexVelezLl I raised the PR #898 , please review it.

**Comment by MisRob** [comment]
Thank you @Sahil-Sinha-11!
